File: solutions/0125-valid-palindrome/Solution.java
public class Solution {
    public boolean isPalindrome(String s) {
        int izq = 0, der = s.length() - 1;
        while (izq <= der) {
            while (izq <= der && !Character.isLetterOrDigit(s.charAt(izq))) izq++;
            while (izq <= der && !Character.isLetterOrDigit(s.charAt(der))) der--;
            if (izq > der) return true;
            if (Character.toLowerCase(s.charAt(izq)) !=
                Character.toLowerCase(s.charAt(der))) return false;
            izq++; der--;
        }
        return true;
    }

    public static void main(String[] args) {
        Solution s = new Solution();
        System.out.println(s.isPalindrome("A man, a plan, a canal: Panama")); // true
        System.out.println(s.isPalindrome("race a car"));                     // false
        System.out.println(s.isPalindrome(" "));                              // true
        System.out.println(s.isPalindrome("0P"));                             // false
    }
}

File: solutions/0125-valid-palindrome/solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    bool isPalindrome(string s) {
        int n = s.size();
        int izq = 0, der = n - 1;
        while (izq <= der) {
            while (izq <= der && !isalnum(static_cast<unsigned char>(s[izq]))) izq++;
            while (izq <= der && !isalnum(static_cast<unsigned char>(s[der]))) der--;
            if (izq > der) return true;
            if (tolower(static_cast<unsigned char>(s[izq])) !=
                tolower(static_cast<unsigned char>(s[der]))) return false;
            izq++; der--;
        }
        return true;
    }
};

int main() {
    Solution s;
    cout << boolalpha;
    cout << s.isPalindrome("A man, a plan, a canal: Panama") << '\n';  // true
    cout << s.isPalindrome("race a car") << '\n';                      // false
    cout << s.isPalindrome(" ") << '\n';                               // true
    cout << s.isPalindrome("0P") << '\n';                              // false
}
